[Fusilli][1/N] Build, test, graph/node/tensor structures, code-cov, logging

sjain-stanford opened the pull request:
This is very much a 1/N -- gets the basic development stuff in place to start authoring the meaty pieces and onboard more collaborators to the project. It includes build, test, graph containers, logging / error reporting and code coverage. In a follow-up (2/N) I plan on enabling CI through GitHub actions. 

# Fusili

Fusili is a C++ Graph API and Frontend to the IREE stack (compiler & runtime), enabling JIT compilation & execution of training and inference graphs. It allows us to expose cuDNN-like primitives backed by IREE code-generated kernels.

:construction: **This project is under active development and APIs may change** :construction:

A side note on naming: 'SharkFuser' is the name of the project (may change as the project evolves), but the library itself is called 'Fusili'. Both are inspired from the term 'Fusions' - a bread-and-butter compiler optimization to improve performance.

![fusili](https://github.com/user-attachments/assets/2da74ab1-ff4f-4b31-8d07-54553f23b984)

Code coverage: 
<img width="861" height="307" alt="image" src="https://github.com/user-attachments/assets/0f2d05a7-df94-4c1b-acf7-5564d328e218" />

sjain-stanford commented:
@stellaraccident since you'd know this best: 
Any suggestions on how the LICENSE/copyright headers should look like, given parts of the code going forward will inevitably look "like cudnn-frontend" when trying to provide a similar API (albeit the backend interfacing pieces will be drastically different). Should we just use MIT license to begin with, or have Apache 2.0 + LLVM exception + append with MIT notice towards the end of it?

Groverkss commented on sharkfuser/include/fusili/attributes.h:
```diff
@@ -0,0 +1,188 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
```
Can we stick to IREE code style? IREE C++ codestyle is derived from LLVM. We use header guards: https://llvm.org/docs/CodingStandards.html#header-guard

sjain-stanford replied:
@Groverkss Header guard changes are in. I'm using LLVM style clang-format. Btw [noticed](https://sourcegraph.com/github.com/iree-org/iree/-/blob/.clang-format?L9) IREE uses Google style clang-format (there's a comment about runtime using Google vs compiler using LLVM style).

stellaraccident commented:
> @stellaraccident since you'd know this best: Any suggestions on how the LICENSE/copyright headers should look like, given parts of the code going forward will inevitably look "like cudnn-frontend" when trying to provide a similar API (albeit the backend interfacing pieces will be drastically different). Should we just use MIT license to begin with, or have Apache 2.0 + LLVM exception + append with MIT notice towards the end of it?

Just stick with Apache like the rest of the code. Unless of course if the code was literally adapted from another source, in which case note that and preserve the original copyright/license as applicable.

stellaraccident commented on sharkfuser/include/fusili.h:
```diff
@@ -0,0 +1,19 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
```
I know everyone hates it, but even in 2025, there are still good reasons to use ifndef style header guards.

sjain-stanford replied:
Done 👍🏻 

I don't like the ifndef style guards particularly either, mainly because they're more verbose and error prone (I'm afraid I might have a typo in the guards somewhere). So I asked Copilot to do it for me :) 

<img width="654" height="374" alt="image" src="https://github.com/user-attachments/assets/b30e5dc0-4fa9-4ad6-96e1-3fee9c436063" />

stellaraccident commented on sharkfuser/include/fusili/attributes/tensor_attributes.h:
```diff
@@ -0,0 +1,190 @@
+#include <variant>
+#include <vector>
+
+#include "../context.h"
```
Please use full paths to include files, relative to the include root.

stellaraccident commented on sharkfuser/include/fusili/graph.h:
```diff
@@ -0,0 +1,144 @@
+#include <memory>
+#include <unordered_set>
+
+#include "attributes/tensor_attributes.h"
```
Full path to includes.

stellaraccident commented on sharkfuser/include/fusili/logging.h:
```diff
@@ -0,0 +1,180 @@
+
+  const std::string &get_message() const { return err_msg; }
+
+  bool is_good() const { return code == error_code_t::OK; }
```
Maybe fewer different words meaning the same thing. Just make this `is_ok()` and if you need the inverse `is_failure()`

sjain-stanford replied:
Done

stellaraccident commented on sharkfuser/include/fusili/logging.h:
```diff
@@ -0,0 +1,180 @@
+
+inline bool &isLoggingEnabled() {
+  static bool log_enabled = []() -> bool {
+    const char *env_val = std::getenv("FUSILI_LOG_INFO");
```
Fine to bootstrap, but we're going to want to be more sophisticated on logging.

sjain-stanford replied:
I agree. Making a mental note to revisit this once we have a better idea of where this fits in the overall stack (i.e., what comes above it and the exact usage model). I should probably look at how IREE / shortfin does this, but also open to ideas if you have. 

Right now we just show the error + line locations and support saving to a log file on disk optionally, but we need a lot more (severity levels, surfacing compiler/runtime errors nicely, capturing stack traces, dumping in a record/json format). 
```
[FUSILI] INFO: Validating graph
[FUSILI] INFO: Validating node Type::Convolution conv_fprop...
[FUSILI] INFO: Inferring properties for node Type::Convolution conv_fprop...
[FUSILI] ERROR: NOT_IMPLEMENTED: Convolution node shape inference not implemented yet; please specify output tensor dimensions: (true) at /home/srajeshk/code/shark-ai/sharkfuser/include/fusili/node/conv_node.h:62
[FUSILI] ERROR: infer_properties_node() at /home/srajeshk/code/shark-ai/sharkfuser/include/fusili/node/node.h:39
[FUSILI] ERROR: sub_node->validate_subtree() at /home/srajeshk/code/shark-ai/sharkfuser/include/fusili/node/node.h:41
[FUSILI] ERROR: validate_subtree() at /home/srajeshk/code/shark-ai/sharkfuser/include/fusili/graph.h:57
```

stellaraccident commented on sharkfuser/include/fusili/node/conv_node.h:
```diff
@@ -0,0 +1,101 @@
+#ifndef FUSILI_NODE_CONV_NODE_H
+#define FUSILI_NODE_CONV_NODE_H
+
+#include "../attributes/conv_attributes.h"
```
Full paths.

stellaraccident commented on sharkfuser/include/fusili/node/conv_node.h:
```diff
@@ -0,0 +1,101 @@
+                             "Convolution node shape inference not implemented "
+                             "yet; please specify output tensor dimensions");
+
+      // auto y_dim_inferred = std::vector<int64_t>(x_dim.size(), 1);
```
Let's just leave the commented/not done yet code out.

stellaraccident commented on sharkfuser/include/fusili/node/node.h:
```diff
@@ -0,0 +1,68 @@
+
+#include <memory>
+
+#include "../context.h"
```
Full path.

stellaraccident reviewed (approved):
lgtm modulo nits

sjain-stanford commented:
Thanks for the review. I've addressed the comments so will go ahead and land. If there's anything more you find in subsequent passes, I'll happily address them in follow-ups.
